content/en/blog/_posts/2022-12-29-scalable-job-tracking-ga/index.md

```diff
@@ -0,0 +1,148 @@
+ignore Pods that don't have a the aforementioned finalizer.
+
+Before you upgrade your cluster to 1.27, we recommend that you verify that there
+are no running Jobs that don't have the annotation. Otherwise, you might observe
+the control plane recreating some Pods. We expect that this shouldn't affect
+any users, as the feature is enabled by default since Kubernetes 1.25.
+
+## What problem does the new implementation solve?
+
+Generally, Kubernetes workload controllers, such as ReplicaSet or StatefulSet,
+rely on the existence of Pods or other objects in the API to determine the
+status of the workload and whether replacements are needed.
+For example, if a Pod that belonged to a ReplicaSet terminates or ceases to
+exist, the ReplicaSet controller needs to create a replacement Pod to satisfy
+the desired number of replicas (`.spec.replicas`).
+
+Since its inception, the Job controller also relied on the existence of Pods in
+the API to track Job status. A Job has [completion](/docs/concepts/workloads/controllers/job/#completion-mode)
+and [failure handling](/docs/concepts/workloads/controllers/job/#handling-pod-and-container-failures)
+policies, requiring the end state of a finished Pod to determine whether to
+create a replacement Pod or mark the Job as completed or failed. As a result,
+the Job controller depended on Pods, even terminated ones, to remain in the API
+in order to keep track of the status.
+
+This dependency made the tracking of Job status unreliable, because Pods can be
+deleted from the API for a number of reasons, including:
+- The garbage collector removing orphan Pods when a Node goes down.
+- The garbage collector removing terminated Pods when they reach a threshold.
+- The Kubernetes scheduler preempting a Pod to accomodate higher priority Pods.
+- The taint manager evicting a Pod that doesn't tolerate a `NoExecute` taint.
+- External controllers, not included as part of Kubernetes, or humans deleting
+  Pods.
+
+### The new implementation
+
+When a controller needs to take an action on objects before they are removed, it
+should add a [finalizer](/docs/concepts/overview/working-with-objects/finalizers/)
+to the objects that it manages.
+A finalizer prevents the objects from being deleted from the API until the
+finalizers are removed. Once the controller is done doing cleanup and accounting
+for the deleted object, it can remove the finalizer from the object and the
+control plane removes the object from the API.
+
+This is what the new Job controller is doing: adding a finalizer during Pod
+creation, and removing it after the Pod terminates and is accounted for in the
+Job status. However, it wasn't that simple.
+
+The main challenge is that there are at least two objects involved: the Pod
+and the Job. While the finalizer lives in the Pod object, the accounting lives
+in the Job object. There is no mechanism to atomically remove the finalizer in
```

Comment:
Do we mean unchangeably remove the finalizer or automatically remove the finalizer here since atomically = unchangeable, indivisible, or infinitesimal?
```suggestion
in the Job object. There is no mechanism to atomically remove the finalizer in
```

Reply:
Ah okay, that clarifies! Thanks :) I confused it with the literal definition.